PMNS_ScalarNSI.cxx

```diff
@@ -0,0 +1,410 @@
+    cout << "Setting reverse order (Eta_" << flvj << flvi << "). " << endl;
+    int temp = flvi;
+    flvi = flvj;
+    flvj = temp;
+  }
+  if(flvi<0 || flvi>2 || flvj < flvi || flvj > 2){
+    cout << "Eta_" << flvi << flvj << " not valid for " << fNumNus;
+    cout << " neutrinos. Returning 0." << endl;
+    return zero;
+  }
+
+  return fEta[flvi][flvj];
+
+}
+/// Set lightest neutrino mass
+void PMNS_ScalarNSI::SetM(double m){
+
+fM=m;
+
+}
+//......................................................................
+///
+/// Set eta_ee parameter
+///
+/// This will check if value is changing to keep track of whether
+/// the eigensystem needs to be recomputed.
+///
+/// @param a - The value of the real parameter eta_ee
+///
+void PMNS_ScalarNSI::SetEta_ee(double a){ SetEta(0,0, a, 0); }
+
+//......................................................................
+///
+/// Set eta_mumu parameter
+///
+/// This will check if value is changing to keep track of whether
+/// the eigensystem needs to be recomputed.
+///
+/// @param a - The value of the real parameter eta_mumu
+///
+void PMNS_ScalarNSI::SetEta_mumu(double a){ SetEta(1,1, a, 0); }
+
+//......................................................................
+///
+/// Set eta_tautau parameter
+///
+/// This will check if value is changing to keep track of whether
+/// the eigensystem needs to be recomputed.
+///
+/// @param a - The value of the real parameter eta_tautau
+///
+void PMNS_ScalarNSI::SetEta_tautau(double a){ SetEta(2,2, a, 0); }
+
+//......................................................................
+
+void PMNS_ScalarNSI::SetEta_emu(double a, double phi){ SetEta(0,1, a, phi); }
+
+//......................................................................
+///
+/// Set eta_mumu parameter
+///
+/// This will check if value is changing to keep track of whether
+/// the eigensystem needs to be recomputed.
+///
+/// @param a - The value of the real parameter eta_mumu
+///
+void PMNS_ScalarNSI::SetEta_mutau(double a, double phi){ SetEta(1,2, a, phi); }
+
+//......................................................................
+///
+/// Set eta_tautau parameter
+///
+/// This will check if value is changing to keep track of whether
+/// the eigensystem needs to be recomputed.
+///
+/// @param a - The value of the real parameter eta_tautau
+///
+void PMNS_ScalarNSI::SetEta_etau(double a, double phi){ SetEta(0,2, a, phi); }
+
+
+////Get lightest neutrino mass
+
+
+double PMNS_ScalarNSI::GetM(){ return fM;}
+/// Set eps_mutau parameter
+///
+/// This will check if value is changing to keep track of whether
+/// the eigensystem needs to be recomputed.
+///
+/// @param a   - The absolute value of the parameter eps_mutau
+///
+/// Build the full Hamiltonian in matter.
+///
+/// Here we divide the mass squared matrix Hms by the 2E
+/// to obtain the vacuum Hamiltonian in eV. Then, the matter
+/// potential is added to the electron component.
+///
+void PMNS_ScalarNSI::UpdateHam()
```

Comment:
Now that I understand a bit better the code, I think we should spend some time rethinking the optimisation of this approach. There is a lot of matrix multiplication here which is slow. In the standard oscillation case we separate the rotation of the Hamiltonian from the adding of matter effects. We want to avoid rebuilding the Hamiltonian from scratch every time we change the energy, because some parts don't change with energy.

Additionally, isn't the M + dM matrix Hermitian? That means you don't have to compute the conjugate-transpose. You can just square it. That would reduce a lot the matrix multiplications here.

Reply:
Here is a rough idea of what you want to do:
```cpp
  BuildHms(); // Modify this to build the M matrix and store in fHam

  // Add the dM matrix
  double sqrtDm31 = sqrt(fabs(fDm[2]));
  for(int i=0; i<fNumNus; i++){
  for(int j=i; j<fNumNus; j++){
    fHam[i][j] += fEta[i][j] * sqrtDm31;
    if(j>i) fHam[j][i] = conj(fHam[i][j]);
  }}

  // Square fHam
  fHam = Dot(fHam, fHam); // Write a matrix multiplication function

  // Divide the Hamiltonian by energy
  double lv_inv = 1 / (2 * kGeV2eV*fEnergy);     // 1/2E in 1/eV
  for(int i=0; i<fNumNus; i++){
  for(int j=0; j<fNumNus; j++){
    fHam[i][j] *= lv_inv;
  }}

  /// Adding matter potential
  double kr2GNe = kK2*M_SQRT2*kGf;
  kr2GNe *= fPath.density * fPath.zoa; // Matter potential in eV
  if(!fIsNuBar) fHam[0][0] += kr2GNe;
  else          fHam[0][0] -= kr2GNe;
```